### PR TITLE
make content layouts verbatim safe

### PR DESCRIPTION
**Pull request by EagleoutIce**
This new variant is based on https://github.com/SoftVarE-Group/SlideTemplate/pull/31, but implements the behavior as defined by @tthuem in https://github.com/SoftVarE-Group/SlideTemplate/issues/26#issuecomment-1237071847.

Furthermore it offers support for vertical alignment (with 't', 'T', 'b' and the default 'c'). 
The current implementation is rather naïve and can be improved easily with a loop (creating all pgfkeys). However this would either a) require another package or b) the usage of lower-level loop-commands. 

Additionally, we can:

1. Issue a deprecation warning whenever someone uses the old commands
2. Fine tune `\halfnext`, `\thirdnext`, and spacing so that the new columns look exactly similar to the previous version (although the new implementation handles spaces etc).
3. Use onlytextwidth for more intuitive width-calculation

### CODE REVIEW

**Comment by Benno2000** [comment]
I think what @tthuem meant was an environment that takes a number of columns and an animation option, for example the replacement of `\leftmiddleorright` could look something like that:

```
\begin{mycolumns}[columns=3, animation=single]
% some content for the first column
\mynextcolumn
% some content for the second column
\mynextcolumn
% some content for the third column
\end{mycolums}
```

Do you think something like that is possible without making the template too complicated? Then we could just leave the original definitions of the content layouts at the end of the template in addition (and mark them as deprecated) and just remove them later.

**Comment by EagleoutIce** [comment]
Okay, i started to work on that on [branch 26-alternate-column-specification](https://github.com/SoftVarE-Group/SlideTemplate/tree/26-alternate-column-specification).

However, the desired names for the different animations are not clear to me.
Why, for example is single intuitively left to right? 

So these are the questions i need answered:
1. What are the names for all desired animations and how should they work?
2. What are potential additional options?
3. What kind of errors should be caught and reported?

**Comment by Benno2000** [comment]
1. We didn't discuss the exact names for the animations yet, `single` was just an example. However I think we need two different types of animations: Something like `single` which corresponds to the "or"-layouts (`\leftorright`, `\leftmiddleorright` etc.) and something like `additional` or `buildup` for the "then"-layouts (`\leftthenright`, `\leftmiddlethenright` etc.). The default would be no animation (corresponding to the "and"-layouts (`\leftandright`, `\leftmiddleandright`)). Something that you mentioned and I didn't consider yet is the direction of the animation (left-to-right vs. right-to-left). Therefore we could maybe use an additional option `direction` which can take the values `lefttoright` and `righttoleft` (with `lefttoright` as the default). 
**@tthuem What are your thoughts on that?**
2. I think `columns` (default: 2), `animation` (default: no animation), and `direction` (default: lefttoright) should be the only options necessary.
3. I have not thought about that yet, I think you can leave the error reports out first off for demonstration purposes.

**Comment by tthuem** [comment]
Having the vertical alignment as option is great. I would have needed this numerous times before. :)

Flo, which errors do you have in mind? I think the environment that I envisioned would always try to render it for any configuration (opposed to giving error messages). For example, if \mynextcolumn is missing, then we simply leave out those columns and center the remaining columns or fill the space before and after every existing column. Not sure what could be useful here.

Most of the time, people want two columns and the animation leftthenright, I guess. So it would be great to have this as easy as possible (e.g., without parameters). For example:

```
\begin{mycolumns}
\mynextcolumn
\end{mycolumns}
```

Parameters:
* `columns` specifies the number of columns
  * `columns=2` is the default, so far I only needed the values 2 and 3 in my slides (4 or more could also be useful in some seldom cases)
* `animation` specifies the overlay behavior (if handout is not activated)
  * `animation=none` (cf. \leftandright)
  * `animation=add` is the default (cf. \leftthenright)
  * `animation=xor` (cf. \leftorright)
* `reverse` or something similar to use the opposite order (cf. \rightthenleft and \rightorleft)

I'm not really happy with the names yet, so proposals are welcome. Also, I'm not sure how easy it is to implement the commands. We could also stick to leftandright, leftorright, rightorleft, leftthenright and rightthenleft. Could be a bit unintuitive what happens for more than two columns.

**Comment by Benno2000** [comment]
That sounds good to me, although I would make `animation=none` the default option. When I use a column environment in a presentation I wouldn't expect it to do animations automatically. I also like the reverse option, it is simpler then using `direction=<...>`
Regarding the names I like your proposals except for `animation=xor`. I would maybe call that `single` or something similar because users of the template with a  non-technical background may not know the meaning of "XOR".

**Comment by tthuem** [comment]
We discussed separate labels for the animations:
* keep (formerly then)
* forget (formerly or)
* animate=none (default, formerly and)

Other ideas but not used:
* inc / add
* single

* include vs exclude
* cumulative vs individuell
* and  vs implies vs xor

**Comment by tthuem** [comment]
New parameter to adjust the width of the columns?

widths={75,25}

**Comment by tthuem** [comment]
@EagleoutIce What is the status here?

I just checked the content, but the new environment is not part of this PR yet. When will you be ready with the new environment? I'm asking because of the release that I would like to publish in the next days.

File: beamerthemeuulm.sty [hunk 1]
```diff
@@ -198,131 +198,259 @@
...
-	\thirdpage{#1}
+\newcommand\leftmiddlethenright[3]{%
+    \thirdpage{#1}%
     \hfill
-    \onslide<2->{\thirdpage{#2}}
+    \onslide<2->{\thirdpage{#2}}%
     \hfill
-    \onslide<3->{\thirdpage{#3}}
+    \onslide<3->{\thirdpage{#2}}%
 }
```
Comment (tthuem) [inline beamerthemeuulm.sty:240]: Why do we want to ignore the third parameter here and duplicate content of the second?

File: beamerthemeuulm.sty [hunk 2]
```diff
@@ -198,131 +198,259 @@
 	\end{minipage}
 }
 
+% corresponds to '\halfpage', however, with the new columns we could use larger values
+% (same for thirdnext)
+\def\halfnext{.489\textwidth}
+% corresponds to '\thirdpage'
+\def\thirdnext{.31\textwidth}
+\def\uulm@cols@error{\PackageError{beamerthemeuulm}{The layout <TODO> you selected does not allow another \string\mynextcolumn\space (page: \thepage, frame: \insertframenumber,\on@line)}{Either remove the command, or choose another Layout}}
```
Comment (tthuem) [inline beamerthemeuulm.sty:206]: Is this a warning or an error? I guess a warning would be enough here.

File: beamerthemeuulm.sty [hunk 3]
```diff
@@ -198,131 +198,259 @@
...
 
-\newcommand{\rightmiddlethenleft}[3]{
-	\onslide<3->{\thirdpage{#1}}
-    \hfill
-    \onslide<2->{\thirdpage{#2}}
+\newcommand\rightmiddlethenleft[3]{%
+    \onslide<3->{\thirdpage{#1}}%
     \hfill
-    \thirdpage{#3}
+    \onslide<2->{\thirdpage{#2}}\hfill\thirdpage{#2}%
```
Comment (tthuem) [inline beamerthemeuulm.sty:245]: Again, duplicated content in second and third column.

File: beamerthemeuulm.sty [hunk 4]
```diff
@@ -198,131 +198,259 @@
...
 }
 
-\newcommand{\leftorright}[2]{ % only works in recording mode
```
Comment (tthuem) [inline beamerthemeuulm.sty:244]: Why did you remove this comment?

File: beamerthemeuulm.sty [hunk 5]
```diff
@@ -198,131 +198,259 @@
...
+\ifrecording
+    \onslide<1>{\halfpage{#1}}%
+    \onslide<2>{\hfill}%
+    \onslide<3>{\halfpage{#2}}%
+\else
+    \leftthenright{#1}{#2}%
+\fi
 }
 
-\newcommand{\rightorleft}[2]{ % only works in recording mode
```
Comment (tthuem) [inline beamerthemeuulm.sty:254]: see above

File: beamerthemeuulm.sty [hunk 6]
```diff
@@ -198,131 +198,259 @@
...
+\ifrecording
+    \onslide<3>{\halfpage{#1}}%
+    \onslide<2>{\hfill}%
+    \onslide<1>{\halfpage{#2}}%
+\else
+    \rightthenleft{#1}{#2}%
+\fi
 }
 
-\newcommand{\leftmiddleorright}[3]{ % only works in recording mode
```
Comment (tthuem) [inline beamerthemeuulm.sty:264]: see above

File: beamerthemeuulm.sty [hunk 7]
```diff
@@ -198,131 +198,259 @@
...
+    \onslide<2>{\hfill}%
+    \onslide<3>{\thirdpage{#2}}%
+    \onslide<4>{\hfill}%
+    \onslide<5>{\thirdpage{#3}}%
+\else
+    \leftmiddlethenright{#1}{#2}{#3}%
+\fi
 }
 
-\newcommand{\rightmiddleorleft}[3]{ % only works in recording mode
```
Comment (tthuem) [inline beamerthemeuulm.sty:276]: see above

File: beamerthemeuulm.sty [hunk 8]
```diff
@@ -198,131 +198,259 @@
...
+    \rightmiddlethenleft{#1}{#2}{#3}%
+\fi
 }
-% ---------------------------------------------
-
 
-% ---------------------------------------------
-% color boxes
-% ---------------------------------------------
-\newcommand{\mydefinition}[2]{
```
Comment (tthuem) [inline beamerthemeuulm.sty:293]: Did you really remove this mydefinition, mydefinitiontight, ...? We still need those.

**Review by tthuem — CHANGES_REQUESTED** [review]
It is rather hard to review this PR as there are too many things changed in the same branch. Also there are merge conflicts to the current release branch (i.e., it is not up-to-date). Maybe it would be easier to redo the changes on the current release on different branches.

There are probably bugs introduced in this PR with respect to the old definitions for columns. The latex code for the new columns and definitions is rather complex and I cannot understand all details. Is it planned to keep all those TODOs in the template?

File: beamerthemeuulm.sty (hunk 2 above)
Comment (EagleoutIce) [reply beamerthemeuulm.sty:206]: Currently i made everything an error. But i can change that to warning.

File: beamerthemeuulm.sty (hunk 1 above)
Comment (EagleoutIce) [reply beamerthemeuulm.sty:240]: Fixed

File: beamerthemeuulm.sty (hunk 4 above)
Comment (EagleoutIce) [reply beamerthemeuulm.sty:244]: I've added them again.

File: beamerthemeuulm.sty (hunk 6 above)
Comment (EagleoutIce) [reply beamerthemeuulm.sty:264]: Same

File: beamerthemeuulm.sty (hunk 7 above)
Comment (EagleoutIce) [reply beamerthemeuulm.sty:276]: Same

File: beamerthemeuulm.sty (hunk 8 above)
Comment (EagleoutIce) [reply beamerthemeuulm.sty:293]: Same

File: beamerthemeuulm.sty (hunk 5 above)
Comment (EagleoutIce) [reply beamerthemeuulm.sty:254]: Same

**Comment by EagleoutIce** [comment]
First of all: I've already dealt with conflicts before your review, do you see any conflicts remain?

No i do not plan to keep those todos. The current state is more like a "do you understand the code"/"what to do to make it understandable".

**Comment by tthuem** [comment]
I tried the new changes but it seems that the default for the parameter `columns` is `columns=1` instead of `columns=2`.
